File: 785 Is Graph Bipartite/785.py
from collections import deque
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        if not graph:
            return False
        n = len(graph)
        colors = [-1 for i in range(n)]
        for index, node in enumerate(graph):
            if node and colors[index] < 0:
                colors[index] = 0
                if not helper(graph, index, colors):
                    return False
        return True


def helper(graph, index, colors):
    queue = deque()
    queue.append(index)
    while queue:
        node = queue.popleft()
        for item in graph[node]:
            if colors[item] < 0:
                colors[item] = colors[node] ^ 1
                queue.append(item)
            elif colors[item] != colors[node] ^ 1:
                return False
    return True

File: 785 Is Graph Bipartite/785.js
/**
 * @param {number[][]} graph
 * @return {boolean}
 */
var isBipartite = function(graph) {
    if (!graph) {
        return false;
    }
    let n = graph.length
    let colors = new Array(n).fill(-1)
    for (let i = 0; i < n; ++i) {
        if (graph[i].length && colors[i] < 0) {
            colors[i] = 0
            if (!helper(graph, i, colors)) {
                return false;
            }
        }
    }
    return true;
};

var helper = function(graph, index, colors) {
    let queue = []
    queue.push(index)
    while (queue.length) {
        let node = queue.shift()
        for (let item of graph[node]) {
            if (colors[item] < 0) {
                colors[item] = colors[node] ^ 1
                queue.push(item)
            } else if (colors[item] != colors[node] ^ 1) {
                return false;
            }
        }
    }
    return true;
}